packages/commerce-sdk-react/src/auth/index.ts

```diff
@@ -652,6 +682,101 @@ class Auth {
+        const organizationId = slasClient.clientConfig.parameters.organizationId
+        const clientId = slasClient.clientConfig.parameters.clientId
+        const siteId = slasClient.clientConfig.parameters.siteId
+        const loginId = credentials.loginId || 'guest'
+        const isGuest = loginId === 'guest'
+        const idpOrigin = isGuest ? 'slas' : 'ecom'
+
+        const url = `${slasClient.clientConfig.proxy}/shopper/auth/v1/organizations/${organizationId}/oauth2/trusted-agent/authorize?${[
+            ...[
+                `client_id=${clientId}`,
+                `channel_id=${siteId}`,
+                `login_id=${loginId}`,
+                `redirect_uri=${this.redirectURI}`,
+                `idp_origin=${idpOrigin}`,
+                `response_type=code`
+            ],
+            ...(!this.clientSecret ? [`code_challenge=${codeChallenge}`] : [])
+        ].join('&')}`
+
+        return {url, codeVerifier}
+    }
+
+    async loginTrustedAgent(credentials: {
+        loginId?: string
+        code: string
+        codeVerifier?: string
+        usid?: string
+        state?: string
+        clientSecret?: string
+    }) {
+        const slasClient = this.client
+        const loginId = credentials.loginId || 'guest'
+        const isGuest = loginId === 'guest'
+        const idpOrigin = isGuest ? 'slas' : 'ecom'
+
+        const optionsToken = {
+            headers: {
+                Authorization: `Bearer ${credentials.code}`
+            },
+            body: {
+                channel_id: slasClient.clientConfig.parameters.siteId,
+                grant_type: 'client_credentials',
+                redirect_uri: this.redirectURI,
+                login_id: loginId,
+                idp_origin: idpOrigin,
+                dnt: 'true',
+                ...(!this.clientSecret && {
+                    client_id: slasClient.clientConfig.parameters.clientId,
+                    code_verifier: credentials.codeVerifier
+                }),
+                ...(credentials.state && {state: credentials.state}),
+                ...(credentials.usid && {usid: credentials.usid})
+            }
+        } as {headers: {[key: string]: string}; body: TrustedAgentTokenRequest}
+
+        // using slas private client
+        if (credentials.clientSecret) {
+            optionsToken.headers._sfdc_client_auth = `Basic ${helpers.stringToBase64(
+                `${slasClient.clientConfig.parameters.clientId}:${credentials.clientSecret}`
+            )}`
+        }
+
+        const token = await slasClient.getTrustedAgentAccessToken(optionsToken)
+        this.handleTokenResponse(token, isGuest)
+
+        return token
+    }
+
+    registerTrustedAgentRefreshHandler(
+        refreshTrustedAgentHandler: (
+            loginId?: string,
+            usid?: string,
+            refresh?: boolean
+        ) => Promise<TokenResponse>
+    ) {
+        this.refreshTrustedAgentHandler = refreshTrustedAgentHandler
+    }
```

Comment:
Registering a constructor option from outside of the class feels like an anti-pattern. Is this because `clearStorage()` isn't exposed to the `useTrustedAgent` hook?

Reply:
yeah this felt out of place to me but unsure what other options we have.  we basically need to inject our popup refresh flow in the `ready()` flow, and that lives in `useTrustedAgent` hook today.  i'll give it some thought.